engine/execution/ingestion/stop_control.go

```diff
@@ -0,0 +1,235 @@
+	s.state = StopControlSet
+
+	s.height = height
+	s.crash = crash
+	s.stopAfterExecuting = flow.ZeroID
+
+	return oldHeight, oldCrash, nil
+}
+
+// GetStopHeight returns:
+//   - height
+//   - crash
+//
+// Values are undefined if they were not previously set
+func (s *StopControl) GetStopHeight() (uint64, bool) {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.height, s.crash
+}
+
+// blockProcessable should be called when new block is processable.
+// It returns boolean indicating if the block should be processed.
+func (s *StopControl) blockProcessable(b *flow.Header) bool {
+
+	s.Lock()
+	defer s.Unlock()
+
+	if s.state == StopControlOff {
+		return true
+	}
+
+	if s.state == StopControlPaused {
+		return false
+	}
+
+	// skips blocks at or above requested stop height
+	if b.Height >= s.height {
+		s.log.Warn().Msgf("Skipping execution of %s at height %d because stop has been requested at height %d", b.ID(), b.Height, s.height)
+		s.log.Debug().Int8("previous_state", int8(s.state)).Int8("new_state", int8(StopControlCommenced)).Msg("StopControl state transition")
+		s.state = StopControlCommenced // if block was skipped, move into commenced state
+		return false
+	}
+
+	return true
+}
+
+// blockFinalized should be called when a block is marked as finalized
+func (s *StopControl) blockFinalized(ctx context.Context, execState state.ReadOnlyExecutionState, h *flow.Header) {
+
+	s.Lock()
+	defer s.Unlock()
+
+	if s.state == StopControlOff || s.state == StopControlPaused {
+		return
+	}
+
+	// Once finalization reached stop height we can be sure no other fork will be valid at this height,
+	// if this block's parent has been executed, we are safe to stop or crash.
+	// This will happen during normal execution, where blocks are executed before they are finalized.
+	// However, it is possible that EN block computation progress can fall behind. In this case,
+	// we want to crash only after the execution reached the stop height.
+	if h.Height == s.height {
+
+		executed, err := state.IsBlockExecuted(ctx, execState, h.ParentID)
+		if err != nil {
+			// any error here would indicate unexpected storage error, so we crash the node
+			s.log.Fatal().Err(err).Str("block_id", h.ID().String()).Msg("failed to check if the block has been executed")
+			return
+		}
+
+		if executed {
+			s.stopExecution()
+		} else {
+			s.stopAfterExecuting = h.ParentID
+		}
+
+	}
+
+}
+
+// blockExecuted should be called after a block has finished execution
+func (s *StopControl) blockExecuted(h *flow.Header) {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.state == StopControlPaused || s.state == StopControlOff {
+		return
+	}
+
+	if s.stopAfterExecuting == h.ID() {
+		// double check. Even if requested stop height has been changed multiple times,
+		// as long as it matches this block we are safe to terminate
+
+		if h.Height == s.height-1 {
+			s.stopExecution()
+		}
```

Comment:
I think the implementation is different from what is described here. 

Is the `else` case expected or unexpected? If is unexpected, better to crash. 

Either way, it's better to log whether the condition is true or not.

```suggestion
		} else {
		     s.Fatal()
		}
```

Reply:
I think it hold true - what I meant was that if you, say, changed stop height from 100 to 200 and back to 100, it should still give the same finalized block. 
I wonder if the second check (height) is necessary. 
I will leave it and add logs in else, but if you think we should simply remove it, I'm find with that